src/main/java/org/spongepowered/api/command/Command.java

```diff
@@ -59,6 +70,38 @@
  * <p>Plugins are free to implement this interface should they prefer to do so.
  * Custom implementations of this class are not required to implement a sane
  * {@link Object#equals(Object)}, but are highly encouraged to do so.</p>
+ *
+ * <p>Commands in Sponge are provided with a {@link Cause}, rather than a direct
+ * source. The cause may contain contexts that provide information about what
+ * they should act upon:</p>
+ *
+ * <ul>
+ *     <li>{@link EventContextKeys#COMMAND_SOURCE}, which may contain a
```

Comment:
@dualspiral 

I disagree with the distinction here. The way I envisioned a Command cause is literally the same as the Event. If we treat the invocation of a Command as an Event then having CommandSource is literally redundant: check the cause for what you're looking for (be it Entity/etc) and move on. Even Command_Entity makes no sense because the Entity would be the DIRECT cause and the other thing who wanted the Entity to be command trigger would be in the cause as well (Such as a Player "executing" a  command as a Creeper). Only use for Command_Entity may be the "centered around" concept.

Reply:
I don't disagree with the removal of the CommandSource, for sure. This is more me thinking through my ideas.

My thought behind the indicated key is to say "this object did not actually cause this command, but use it as the source". I could just put the source into the root of the cause, sure, but then, the cause doesn't really show the actual cause - it becomes some sort of fake cause, really. It depends on how pure you want the cause to really be. My other thought was mostly on permissions, mostly going to look for the first `Subject` in the source, but the idea is that again it can be overridden by the specific event context.

When I get this properly running, I'm gonna then investigate the removal of `CommandSource`. However, are we going to need to replace it by a `CommandInvoker` which is likely to be little more than extended `MessageReciever`/`Subject` interface? I'm effectively using the `CommandSource` as an invoker as it is - is it up to the plugin to decide this and not give any hint as to who the invoker is? I'm already leaning this way as I type this - though I'd want to offer a targeted executor (like I've added) that filters out the intended target - which might be a combination of interfaces (such as `Subject` and `MessageReceiver`).

Reply:
add the "source" as a context then?

Reply:
Well I would still view it as the direct cause in that example with the Player in the cause too. If someone wants to do something special if a command was invoked by someone not of their own choosing (Player makes Creeper execute command as an example) then they check to see if the Creeper is not the only thing in the cause perhaps?

Reply:
It should just stay easy to get whatever we want to send messages to in the command.
Searching the cause is fine as an option if you need more control.

Reply:
One thought I had it is to go futher on the high-level command treatment and use a subset of the event filters to do some of this (and move the cause into the `CommandContext`).

```java
/**
 * The default command executor method - always present, a fallback should always be offered
 */
public CommandResult execute(CommandContext context);

/**
 * A filtered context, overrides the default
 */
@Executor // used to mark that we want to generate a handler via ASM
public CommandResult execute(CommandContext context, @First Player player);

/**
 * A filtered context, overrides the default, caused by an entity but is attributed to the console
 */
@Executor
public CommandResult execute(CommandContext context, @First ConsoleSource console, @Root Entity entity);

/**
 * A filtered context, overrides the default, caused by an creeper but spawned by the Player - Zidane's example
 */
@Executor
public CommandResult execute(CommandContext context, @First Player player, @Root Creeper creeper);
```

The possibilities of this are then quite flexible. My thought is that if you want a command to be run as someone else, but that someone else isn't the cause, should they go in the cause? I'm honestly leaning towards adding a method on the command manager `process(String command, Object source, CommandTarget target)` that puts the entries on the cause stack correctly so that we can standardise this.

Finally, I have concerns about permission checks, should that default to the first `Subject` in the cause stack? I think that's the trickiest part of the system for me - we just need to be consistent as to what we do here, and document it _very_ clearly.

Reply:
I dislike the impression above that Cause is within Context when in reality they are two separate things. As a counter point:

``` java
/**
 * The default command executor method - always present, a fallback should always be offered
 */
public CommandResult execute(Cause cause, CommandContext context);
```

...is consistent with the event API and makes it clear to consumers that you have your cause of the command and your contextual data.

Thoughts?

Reply:
@Faithcaio 

Having the Cause object means its just as easy and the same way you would do in an event lol. Use event filters and send it to the Player.

To continue my previous post, the permission part really would be the only thing here.

Reply:
The unifying usage of the sender of a command is to send messages back to it, almost every command every written will do this and hiding the sender in a cause is a needless annoyance placed on every plugin for no gain.

Reply:
In Addition, if we want people to correctly use permissions etc from the proxy, or message the proxy, whilst acting on the target of the command we need to be really explicit and careful. Otherwise people will default to incorrect behaviour.

Reply:
For a plugin _calling_ a command, yes, but for a plugin effectively _listening_ to a command, no - it's not consistent. The Event API exposes the cause within the event object itself, and my thought here was that the `CommandContext` is, in effect, an event object that you can get the cause out of - our event objects all contain `getCause()`. We could effectively split the command system into two if we wanted to, registering a command and listening to an event that's called when a command is invoked if we really want to - but to be honest, I'm not advocating that we do for a second.

I've not had much time to think about all the different possibilities, but there seems to be opposition to doing away with what is currently the `CommandSource` object - and that's understandable. In 1.13, as far as I can see, command invocation is effectively split into a two party situation:

* The *source* of the command invocation - that would be represented by the `Cause`.
* The *owner* of the command invocation - that would be our traditional `CommandSource`, and may or may not be in the cause (but usally will be)

I would expect no-one is against exposing the `Cause` and encouraging plugins to use that to add additional context and value to their commands. I think that we've got the problem where everyone has a different interpretation of how the `Cause` should integrate with the commands. My concern has been permissions - if a non-Subject calls a command, whose permissions are checked - and preserving what I see as the purity of the cause stack - namely someone causing a command to run as another player without the other player being involved - a `/sudo` like command, if you will. Granted, it's niche, but a plugin can then see that the source and the owner are two different things.

Is this an artifact of the current command system where only select objects can execute a command? Maybe, but there is a valid concern here where it's not obvious who exactly is running a command as whom - hence why I have added the various event context keys to try to allow for this to be more explicit when a plugin feels it needs to be.

Ultimately, if all this just makes things more confusing for consumers, or hides what was previously obvious behaviour, then we're doing it wrong. Getting such comments back from Sponge core developers puts red flags in my mind as it is - something isn't sitting well and we need to stamp that out. Going futher, doing recent research, I notice some plugin developers aren't even comfortable with the idea of event filters, and aren't sure when to use `@First` and `@Root` in an event method. We shouldn't need to hold their hand, but we also can't make it too confusing so that no-one wants to use Sponge. Indeed, I am already aware of a few developers not using the arguments system because they don't understand and instead doing things in the Bukkit way by using `GenericArguments.remainingJoinedStrings()` or whatever it is.

We need to look at how Mojang will select a command target based on the source that executes it, and try ot build our systems with that in mind.